### ISSUE TITLE
CodeSnippet: add new prop to select text from different elements instead of code 

### ISSUE THREAD

**Issue by labmate** [carbon-design-system/carbon#9238]
### Summary

- We would like to format the code inside the codesnippet component with spans to be more readable in case of a big log, but it gets copied as [object][object]... 
The problem is if we remove the formatting, then it looks a little bit unreadable. 

### Must-have functionality
We would like to pass an extra selector where the text should be copied from. If it is not given, then fallback to the default code element.

### Specific timeline issues / requests
We hope that this could be done within a 2 week period.

### Available extra resources
- we can collaborate to make this work if needed and open PR.

**Comment by emyarod** [comment]
can you provide some more details about your use case of formatting the code using spans (feel free to create a [reduced test case in Code Sandbox](https://codesandbox.io/s/github/carbon-design-system/carbon/tree/main/packages/react/examples/codesandbox) for easier testing and debugging)? the component currently expects children of type `string`

**Comment by labmate** [comment]
@emyarod 
Yeah sure. We have to display logs which are commonly has 40+ line of codes.
We get the log as an array of objects. Each line is a property on the object. Therefor we use map function to iterate through the object and display the current line. 
With this case we can highlight the important parts of the code. 
Example code:
`const codes = logs.map((item, index) => (
    <span
      style={{
        color: `rgb(${item.fg})`,
        backgroundColor: `rgb(${item.bg})`,
      }}
    >
      {item.content}
    </span>
  ));`
`<CodeSnippet full type="multi" >
   {codes}
 </CodeSnippet>
`

What we would like to achieve is in this case, to store the code in a hidden field as string, and pass the selector of it to the codesnippet as targetSelector where the copied text comes from. Yeah I know this would be a duplicated content, but for the readability we need this solution.
